_posts/2017-11-29-Express-Reloadable-Update.markdown

```diff
@@ -0,0 +1,47 @@
+---
+layout: post
+title: "Optimizing Linked Package Development in Express.js Apps"
+date: 2017-11-29 14:18
+comments: true
+author: chris
+categories: [Express, Node, DX, @artsy/express-reloadable]
+---
+
+In [Modernizing Force](artsy.github.io/blog/2017/09/05/Modernizing-Force/) we discussed some of the tools we've been working with to modernize [Artsy.net](https://www.artsy.net/)'s development environment, from introducing Babel and React to the creation of [@artsy/stitch](https://github.com/artsy/stitch). Increasing overall development speed was another aim, and to that end we released [@artsy/express-reloadable](https://github.com/artsy/express-reloadable) which automatically hot-swaps Express.js code without the restart. In this post I'd like to cover some of the issues we've faced since then, and in particular our solution to library code-sharing in Express apps.
+
+<!-- more -->
+
+It's common to share NPM packages across projects, and oftentimes packages are developed in parallel. Package `A` depends on `B`, but `B` has a bug and you don't want to have to republish (and reinstall) the package in order to see changes made locally. `yarn link` (or `npm link`) was developed for instances like this and while it works great for stop and start processes where boot time is quick, it falls short if the development environment takes a while to load. In UI-rich environments like [Positron](https://github.com/artsy/positron) (our Editorial "Writer" app) and [Force](https://github.com/artsy/force), each boot would come at a high time-cost due to upfront compilation of assets. Tools like [nodemon](https://github.com/remy/nodemon) would automatically stop and start our server process when assets changed but that still didn't alleviate slow iteration times.
+
+To recap from a [previous post](artsy.github.io/blog/2017/09/05/Modernizing-Force/), [@artsy/express-reloadable](https://github.com/artsy/express-reloadable) allows devs to immediately see changes to running Express.js app code:
+
+```javascript
+import express from 'express'
+import { createReloadable, isDevelopment } from '@artsy/express-reloadable'
+
+const app = express()
+
+if (isDevelopment) {
+  const mountAndReload = createReloadable(app, require)
+  mountAndReload('./api')
+}
+```
+
+Changes made within the `api` folder are detected and hot-swapped in, and all that's required is a new http request. However, we found an exception while building out Artsy's new [editorial pages](https://www.artsy.net/article/artsy-editorial-midwest-made-artists-mike-kelley-jim-shaw), which involved [sharing Reaction components](https://github.com/artsy/positron/blob/master/client/apps/edit/components/content/section_tool/index.jsx#L11) between Positron [and Force](https://github.com/artsy/force/blob/master/desktop/apps/article/components/InfiniteScrollArticle.js#L9). Even though we ran `yarn link @artsy/reaction` in each consumer app, changes would not be detected and so we had to do a full restart.
```

Comment:
"sharing Reaction components" is somewhat making an assumption that you know that reaction is a set of React components. Maybe "sharing React components from ours shared module React"?

Reply:
This sounds a bit like a bug to me, tbh. Maybe the solution is to follow su links and watch those files instead, rather than having to add manual configuration?

Reply:
@damassi I see that I was not ‘watching’ the express-reloadable repo, so missed [the issue](https://github.com/artsy/express-reloadable/issues/2) and [the PR](https://github.com/artsy/express-reloadable/pull/3) :/ Will reply there.

Reply:
https://github.com/artsy/express-reloadable/issues/2#issuecomment-347210668

Reply:
If anyone else would like to chime in: https://github.com/artsy/express-reloadable/issues/2#issuecomment-347315292